src/modules/screen_plugins/wl.c

```diff
@@ -1,200 +1,252 @@
+        free(output->name);
+        free(output);
+    }
+    if (display->wl_registry) {
+        wl_registry_destroy(display->wl_registry);
+    }
+    if (display->screencopy_manager) {
+        zwlr_screencopy_manager_v1_destroy(display->screencopy_manager);
     }
-    if (buffer.data) {
-        munmap(buffer.data, (size_t)buffer.height * buffer.stride);
-        buffer.data = NULL;
+}
+
+static int get_frame_brightness(const char *id, const char *env) {
+    struct cl_display display = {};
+    struct cl_output *output;
+    int ret = 0;
+
+    display.wl_display = fetch_wl_display(id, env);
+
+    if (display.wl_display == NULL) {
+        fprintf(stderr, "display error\n");
+        ret = WRONG_PLUGIN;
+        goto err;
     }
-    if (output) {
-        wl_output_destroy(output);
-        output = NULL;
+    wl_list_init(&display.outputs);
+
+    display.wl_registry = wl_display_get_registry(display.wl_display);
+    wl_registry_add_listener(display.wl_registry, &registry_listener, &display);
+    wl_display_roundtrip(display.wl_display);
+    if (display.shm == NULL) {
+        fprintf(stderr, "Compositor is missing wl_shm\n");
+        ret = COMPOSITOR_NO_PROTOCOL;
+        goto err;
     }
-     if (registry) {
-        wl_registry_destroy(registry);
+    if (display.screencopy_manager == NULL) {
+        fprintf(stderr, "Compositor is screencopy manager\n");
+        ret = COMPOSITOR_NO_PROTOCOL;
+        goto err;
     }
-    if (frame) {
-        zwlr_screencopy_frame_v1_destroy(frame);
+    if (wl_list_empty(&display.outputs)) {
+        fprintf(stderr, "No outputs available\n");
+        ret = UNSUPPORTED;
+        goto err;
     }
-    if (shm) {
-        wl_shm_destroy(shm);
+    int sum = 0;
+    wl_list_for_each(output, &display.outputs, link) {
+        struct cl_frame *frame = calloc(1, sizeof(struct cl_frame));
+        output->frame = frame;
+        output->screencopy_frame = zwlr_screencopy_manager_v1_capture_output(
+            display.screencopy_manager, 0, output->wl_output);
+        zwlr_screencopy_frame_v1_add_listener(
+            output->screencopy_frame, &screencopy_frame_listener, output);
+
+        while (!output->frame->copy_done && !output->frame->copy_err &&
+               wl_display_dispatch(display.wl_display) != -1) {
+            // This space is intentionally left blank
+        }
+        if (output->frame->copy_done) {
+            output->frame->brightness = rgb_frame_brightness(
+                output->buffer->shm_data, output->frame->width,
+                output->frame->height, output->frame->stride);
+            sum += output->frame->brightness;
+        }
+    }
+    if (sum > 0) {
+        sum = sum / wl_list_length(&display.outputs);
     }
-    if (screencopy_manager) {
-        zwlr_screencopy_manager_v1_destroy(screencopy_manager);
+err:
+    wl_cleanup(&display);
+    if (ret != 0) {
+        return ret;
     }
-     // NOTE: dpy is disconnected on program exit to workaround
```

Comment:
Can you report this note in the `wl_cleanup` method? I want it to stick around :)

Reply:
of course! I'm not sure why I removed it in the first place. I think because I had used a different display connection method.